### PR TITLE
Setup OSX.1100.ARM64.* helix queues

### PR DESCRIPTION
**Pull request by sdmaclea**
/cc @mangod9

### CODE REVIEW

File: eng/pipelines/libraries/helix-queues-setup.yml
Comment (mangod9) [inline eng/pipelines/libraries/helix-queues-setup.yml:89]: does this need to be scoped to `outerloop` or something so it doesnt trigger for PR jobs just yet?

Comment (safern) [reply eng/pipelines/libraries/helix-queues-setup.yml:89]: or if we want it on CI use `eq(parameters.jobParameters.isFullMatrix, true)`

Comment (sdmaclea) [reply eng/pipelines/libraries/helix-queues-setup.yml:89]: So this doesn't trigger jobs per se.  It should define the helix queues for a platform.

I am not sure this is the appropriate place to limit the usage of the platform queues.  When I had `OSX_arm64` in the `platformGroup` `all` there were flames everywhere because the helix-queues weren't defined.

I suspect we need to gate the jobs elsewhere.  Possibly in the platform-matrix.  Perhaps `OSX_arm64` should only be in the `all` group during `ci` runs.

Comment (safern) [reply eng/pipelines/libraries/helix-queues-setup.yml:89]: Yeah we usually limit the usage of helix queues for a platform here. However, I just realized even if we make this change at the moment, tests wont run in that queue as the jobs that submit the test jobs don't have `OSX_arm64` in their platform list. I think we should make this addition to the helix queue list once we actually add test runs for `OSX_arm64`. 

When we do that we should first know if we have machine capacity to enable it for PRs, or only for runtime tests and then for libraries, and so on. 

cc: @steveisok as he is experimenting with this as well.

Comment (sdmaclea) [reply eng/pipelines/libraries/helix-queues-setup.yml:89]: OK.  I'll enable the CI tests in this PR as well.

Initially I expect the runs to have a lot of intermittent failures, so I don't want to enable them for PRs even if we have capacity.

Are we OK with osx-arm64 CI failures for a while while we collect lists of failing tests.  Or should I set something like `shouldContinueOnError`?

Comment (mangod9) [reply eng/pipelines/libraries/helix-queues-setup.yml:89]: Yeah enabling CI tests should be fine. Do failures in a `CI` job cause any consequences like the build not getting published to nuget feeds etc?

Comment (safern) [reply eng/pipelines/libraries/helix-queues-setup.yml:89]: > Are we OK with osx-arm64 CI failures for a while while we collect lists of failing tests. Or should I set something like shouldContinueOnError?

You could instead add them to `runtime-staging` pipeline which it's intent is to add new platforms that are not stable. 

Another approach is disable all failures without merging. However in the `runtime` pipeline we can't merge with any failures as that pipeline we track it's pass rate. ContinueOnError might be another approach if adding to `runtime-staging` is too complicated.

Note that it seems like the Helix M1 queues don't have any machines yet. At least from what I can see on the kusto queries.

**Comment by sdmaclea** [comment]
The current version should enable `osx-arm64` test for CI runs.  Is there a way to trigger them before merging?  Or should this be done in a infra branch.

**Comment by safern** [comment]
> The current version should enable osx-arm64 test for CI runs. Is there a way to trigger them before merging? Or should this be done in a infra branch.

You can queue a manual build from your branch.

**Comment by sdmaclea** [comment]
> Helix M1 queues don't have any machines yet

OK I'll hold off testing till the queues are populated.

> However in the runtime pipeline we can't merge with any failures as that pipeline we track it's pass rate

I can disable the unstable tests before merging.

File: eng/pipelines/libraries/helix-queues-setup.yml
Comment (marek-safar) [reply eng/pipelines/libraries/helix-queues-setup.yml:89]: /cc @steveisok

**Comment by mangod9** [comment]
Hey @sdmaclea, the queues should be hydrated now, could you please try running tests on them?

**Comment by steveisok** [comment]
I ran libraries tests on https://github.com/dotnet/runtime/pull/47864

Looks like there are various python problems.  Not sure if we're going to hit the same things:

https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-47864-merge-39d501d2b0984d9ea6/System.Buffers.Tests/console.fba4d13d.log?sv=2019-07-07&se=2021-03-15T18%3A41%3A56Z&sr=c&sp=rl&sig=Vl6KbD0L0CdGQOe5CJ%2F42cXOHHO8n090lOIm9ZVaVAw%3D

**Comment by mangod9** [comment]
Thanks @steveisok for the info. Is anyone looking into the python script issues? I do see a mention of script issues being worked on here: https://github.com/dotnet/core-eng/issues/11937#issuecomment-780183832. @parose1 is this something you are still looking into?

**Comment by parose1** [comment]
@mangod9 The python issue I was looking into was blocking the helix script from completing due to a conflict with the python version homebrew installs. This prevented the machines from being added to helix. The solution was to just uninstall the homebrew python and let the helix script install it instead, which allowed the script to complete.

**Comment by sdmaclea** [comment]
@parose1   Can. you provide more info?  What version of python is installed on the machines now?  The initial plan was to have the universal binary version of python 3.9.1 installed.  Is that still installed and on the default PATH?

**Comment by parose1** [comment]
@sdmaclea Running "python3 --version" I see python 3.9.1 is installed on the machines. The only difference is that the helix script installs it instead of homebrew. @ulisesh made the suggestion to uninstall the homebrew version and let the helix script install to get me unblocked, so I could add these into the helix queue, so he may know more on the reasoning as to why the homebrew version was causing issues with the helix script.

**Comment by ulisesh** [comment]
> @parose1 Can. you provide more info? What version of python is installed on the machines now? The initial plan was to have the universal binary version of python 3.9.1 installed. Is that still installed and on the default PATH?

The default python in the PATH has Intel binaries. We had problems using Python with universal binaries because there were some package (like cryptography) that don't support M1

@sdmaclea can you please help us understand how having Intel based Python affects your scenario?

Talking about https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-47864-merge-39d501d2b0984d9ea6/System.Buffers.Tests/console.fba4d13d.log?sv=2019-07-07&se=2021-03-15T18%3A41%3A56Z&sr=c&sp=rl&sig=Vl6KbD0L0CdGQOe5CJ%2F42cXOHHO8n090lOIm9ZVaVAw%3D it looks like the test reporter is having problems running on Python 3.9 and the recommendation [here ](https://stackoverflow.com/questions/65640476/pip3-on-python3-9-fails-on-htmlparser-object-has-no-attribute-unescape) is to upgrade **setuptools**. @parose1 could you run **python3 -m pip install --upgrade setuptools** in all the machines?

**Comment by parose1** [comment]
Talked with @ulisesh on teams and changed the command a bit to "sudo -H python3 -m pip install --no-user --upgrade setuptools"

I completed running that on all the machines in the osx.1100.arm64.open and osx.1100.arm64 helix queues. Let me know if you need anything else, thanks.

**Comment by sdmaclea** [comment]
Looks like the M1 helix queue issue is resolved.

I triggered a jitstress run, but it couldn't run on M1 queuse because it wanted to use .NET 5 SDK osx-arm64 to run the tests.

https://dev.azure.com/dnceng/public/_build/results?buildId=1010529&view=results

So this seems at least partially blocked on #48462.

**Comment by mangod9** [comment]
Ok getting closer, hoping the .net 6 integration into runtime get things working.

**Comment by sdmaclea** [comment]
Outerloop finished with some test errors.  Some were expected, some look like build configuration errors, and some were new.  

- There were a handful of pal errors.  
- Two new other tests failed with the same assert as #47294
- A few file not found issues

Jitstress run. is also making progress.

The libraries outerloop run failed early with a configuration error.

**Comment by mangod9** [comment]
ok nice. Based on this is there an automated nightly pipeline which we can enable for macos arm64?

**Review by mangod9 — APPROVED** [review]

**Comment by mangod9** [comment]
Are tests actually running on osx-arm64? I am noticing this failure in the jitstress runs:

```
Microsoft.DotNet.XUnitConsoleRunner v2.5.0 (64-bit .NET 6.0.0-preview.2.21117.5)
  Discovering: tracing.eventcounter.XUnitWrapper
  Discovered:  tracing.eventcounter.XUnitWrapper
  Starting:    tracing.eventcounter.XUnitWrapper
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.<>c.<.cctor>b__17_0(Object state)
   at System.Threading.Tasks.AwaitTaskContinuation.RunCallback(ContextCallback callback, Object state, Task& currentTask)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__140_1(Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
```

**Comment by sdmaclea** [comment]
Yes.  That test suite had run in earlier runs.  It looks like a spurious failure in that leg of that particular run.

If you remove the failed+aborted filter in the `test` tab, you can see all the tests run and passed.

I looked at outerloop and some of the other jitstress legs/runs and they look fine.

I think this is close to ready to merge.

@hoyosjs @safern Any objections?

**Comment by sdmaclea** [comment]
I wanted to trigger a libraries-outer fullMatrix run on this before merging. The one triggered through `azp run` doesn't trigger the osx-arm64 legs, because it is not full mattrix. 

The `ilasm` legs have a similar issue since the PR trigger limits them to PR which change ilasm/ildasm source code.

**Comment by hoyosjs** [comment]
I'll take a look at this later.

**Comment by sdmaclea** [comment]
Triggered libraries outerloop here https://dev.azure.com/dnceng/public/_build/results?buildId=1014515&view=results

**Comment by sdmaclea** [comment]
I triggered the ilasm test here.

**Comment by sdmaclea** [comment]
The `coreclr-release-outerloop-nightly` has been broken on all other platform for a while.  I didn't retrigger it.

The various other jitstress runs I didn't retrigger.  The earlier runs were very similar to each other.

**Comment by sdmaclea** [comment]
I plan to merge this today.

**Comment by ViktorHofer** [comment]
@safern can you please take a look at the yml changes in particular?

**Review by safern — COMMENTED** [review]
It seems like libraries tests are only going to be run for outerloop... do we want to add a libraries test run for this on innerloop libraries tests that run only when IsFullMatrix == true?

File: eng/pipelines/libraries/helix-queues-setup.yml
Comment (safern) [inline eng/pipelines/libraries/helix-queues-setup.yml:90]: I know we discussed this earlier, but I believe we should remove this condition for libraries and then instead condition the test runs in the yml entry point?

**Comment by sdmaclea** [comment]
I plan several other PR to gradually increase coverage.  I am going to merge this so we begin getting test coverage.  I'll put up a PR to for library tests tomorrow.

**Pull request merged**
